cardano-node-emulator/test/Plutus/Examples/Escrow.hs

```diff
@@ -0,0 +1,485 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -g -fplugin-opt PlutusTx.Plugin:coverage-all #-}
+
+-- | A general-purpose escrow contract in Plutus
+module Plutus.Examples.Escrow (
+  -- $escrow
+  Escrow,
+  EscrowError (..),
+  AsEscrowError (..),
+  EscrowParams (..),
+  EscrowTarget (..),
+  payToScriptTarget,
+  payToPaymentPubKeyTarget,
+  targetTotal,
+  payRedeemRefund,
+  typedValidator,
+
+  -- * Actions
+  pay,
+  redeem,
+  refund,
+  badRefund,
+  RedeemFailReason (..),
+  RedeemSuccess (..),
+  RefundSuccess (..),
+
+  -- * Exposed for test endpoints
+  Action (..),
+
+  -- * Coverage
+  covIdx,
+) where
+
+import Control.Lens (makeClassyPrisms)
+import Control.Monad (void)
+import Control.Monad.Except (catchError, throwError)
+import Control.Monad.RWS.Class (asks)
+import Data.Map qualified as Map
+
+import Cardano.Api qualified as C
+import Cardano.Api.Shelley qualified as C
+import PlutusTx (ToData)
+import PlutusTx qualified
+import PlutusTx.Code (getCovIdx)
+import PlutusTx.Coverage (CoverageIndex)
+import PlutusTx.Prelude ()
+import PlutusTx.Prelude qualified as PlutusTx
+
+import Cardano.Node.Emulator qualified as E
+import Cardano.Node.Emulator.Internal.Node (
+  SlotConfig,
+  pSlotConfig,
+  posixTimeRangeToContainedSlotRange,
+ )
+import Cardano.Node.Emulator.Test (testnet)
+import Data.Maybe (fromJust)
+import Ledger (POSIXTime, PaymentPubKeyHash (unPaymentPubKeyHash), TxId, getCardanoTxId)
+import Ledger qualified
+import Ledger.Tx.CardanoAPI qualified as C
+import Ledger.Typed.Scripts (validatorCardanoAddress)
+import Ledger.Typed.Scripts qualified as Scripts
+import Plutus.Script.Utils.Scripts (ValidatorHash, datumHash)
+import Plutus.Script.Utils.V2.Contexts (
+  ScriptContext (ScriptContext, scriptContextTxInfo),
+  TxInfo,
+  scriptOutputsAt,
+  txInfoValidRange,
+  txSignedBy,
+ )
+import Plutus.Script.Utils.V2.Typed.Scripts qualified as V2
+import Plutus.Script.Utils.Value (Value, geq, lt)
+import PlutusLedgerApi.V1.Interval qualified as Interval
+import PlutusLedgerApi.V2 (Datum (Datum))
+import PlutusLedgerApi.V2.Contexts (valuePaidTo)
+import PlutusLedgerApi.V2.Tx (OutputDatum (OutputDatum))
+
+data RedeemFailReason = DeadlinePassed | NotEnoughFundsAtAddress
+  deriving stock (Eq, Show)
+
+data EscrowError
+  = RedeemFailed RedeemFailReason
+  | RefundFailed
+  deriving stock (Show)
+
+makeClassyPrisms ''EscrowError
+
+{- $escrow
+The escrow contract implements the exchange of value between multiple
+parties. It is defined by a list of targets (public keys and script
+addresses, each associated with a value). It works similar to the
+crowdfunding contract in that the contributions can be made independently,
```

Comment:
Do we have a crowdfunding example already or do we need to revisit the description?

Reply:
We will have one.